In [1]:
# Importing libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split, cross_val_score
from patsy import dmatrices
from sklearn import metrics
import statsmodels.api as sm


In [2]:
# Read data
dta = sm.datasets.fair.load_pandas().data

In [3]:
dta.head()

rate_marriage   age  yrs_married  children  religious  educ  occupation  \
0            3.0  32.0          9.0       3.0        3.0  17.0         2.0   
1            3.0  27.0         13.0       3.0        1.0  14.0         3.0   
2            4.0  22.0          2.5       0.0        1.0  16.0         3.0   
3            4.0  37.0         16.5       4.0        3.0  16.0         5.0   
4            5.0  27.0          9.0       1.0        1.0  14.0         3.0   

   occupation_husb   affairs  
0              5.0  0.111111  
1              4.0  3.230769  
2              5.0  1.400000  
3              5.0  0.727273  
4              4.0  4.666666

In [4]:
# check any null value present or not
dta.isnull().sum()

rate_marriage      0
age                0
yrs_married        0
children           0
religious          0
educ               0
occupation         0
occupation_husb    0
affairs            0
dtype: int64

In [6]:
# understanding data type
dta.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6366 entries, 0 to 6365
Data columns (total 9 columns):
rate_marriage      6366 non-null float64
age                6366 non-null float64
yrs_married        6366 non-null float64
children           6366 non-null float64
religious          6366 non-null float64
educ               6366 non-null float64
occupation         6366 non-null float64
occupation_husb    6366 non-null float64
affairs            6366 non-null int32
dtypes: float64(8), int32(1)
memory usage: 422.8 KB


In [5]:
# Converting affairs into categorical
# add "affair" column: 1 represents having affairs, 0 represents not
dta['affairs'] = (dta['affairs']>0).astype(int)
dta.head()

rate_marriage   age  yrs_married  children  religious  educ  occupation  \
0            3.0  32.0          9.0       3.0        3.0  17.0         2.0   
1            3.0  27.0         13.0       3.0        1.0  14.0         3.0   
2            4.0  22.0          2.5       0.0        1.0  16.0         3.0   
3            4.0  37.0         16.5       4.0        3.0  16.0         5.0   
4            5.0  27.0          9.0       1.0        1.0  14.0         3.0   

   occupation_husb  affairs  
0              5.0        1  
1              4.0        1  
2              5.0        1  
3              5.0        1  
4              4.0        1

In [7]:
dta.columns

Index(['rate_marriage', 'age', 'yrs_married', 'children', 'religious', 'educ',
       'occupation', 'occupation_husb', 'affairs'],
      dtype='object')

In [11]:
# create dataframes with an intercept column and dummy variables for occupation and occupation_husb
y, x = dmatrices('affairs ~ rate_marriage + age + yrs_married + children + religious + educ + C(occupation) + C(occupation_husb)',dta, return_type="dataframe")

In [12]:
x.columns

Index(['Intercept', 'C(occupation)[T.2.0]', 'C(occupation)[T.3.0]',
       'C(occupation)[T.4.0]', 'C(occupation)[T.5.0]', 'C(occupation)[T.6.0]',
       'C(occupation_husb)[T.2.0]', 'C(occupation_husb)[T.3.0]',
       'C(occupation_husb)[T.4.0]', 'C(occupation_husb)[T.5.0]',
       'C(occupation_husb)[T.6.0]', 'rate_marriage', 'age', 'yrs_married',
       'children', 'religious', 'educ'],
      dtype='object')

In [13]:
y.columns

Index(['affairs'], dtype='object')

In [14]:
x = x.rename(columns={'C(occupation)[T.2.0]':'occ_2',
                        'C(occupation)[T.3.0]':'occ_3',
                        'C(occupation)[T.4.0]':'occ_4',
                        'C(occupation)[T.5.0]':'occ_5',
                        'C(occupation)[T.6.0]':'occ_6',
                        'C(occupation_husb)[T.2.0]':'occ_husb_2',
                        'C(occupation_husb)[T.3.0]':'occ_husb_3',
                        'C(occupation_husb)[T.4.0]':'occ_husb_4',
                        'C(occupation_husb)[T.5.0]':'occ_husb_5',
                        'C(occupation_husb)[T.6.0]':'occ_husb_6'})
x.head(2)

Intercept  occ_2  occ_3  occ_4  occ_5  occ_6  occ_husb_2  occ_husb_3  \
0        1.0    1.0    0.0    0.0    0.0    0.0         0.0         0.0   
1        1.0    0.0    1.0    0.0    0.0    0.0         0.0         0.0   

   occ_husb_4  occ_husb_5  occ_husb_6  rate_marriage   age  yrs_married  \
0         0.0         1.0         0.0            3.0  32.0          9.0   
1         1.0         0.0         0.0            3.0  27.0         13.0   

   children  religious  educ  
0       3.0        3.0  17.0  
1       3.0        1.0  14.0

In [15]:
# flatten y into a 1-D array
y = np.ravel(y)
y

array([1., 1., 1., ..., 0., 0., 0.])

### Apply Logistic Regression

In [16]:
# fit model
model = LogisticRegression()
model = model.fit(x, y)

In [17]:
# Model accuracy
model.score(x, y)

0.7258875274897895

In [18]:
# what percentage had affairs?
y.mean()

0.3224945020420987

In [19]:
# examine the coefficients
for el in zip(x.columns, np.transpose(model.coef_).tolist()):
    print(el)

('Intercept', [1.4898360570592706])
('occ_2', [0.1880663965967619])
('occ_3', [0.4989481225612341])
('occ_4', [0.2506681436923961])
('occ_5', [0.8390080011190776])
('occ_6', [0.8339082482822548])
('occ_husb_2', [0.19063622424850235])
('occ_husb_3', [0.29783290491447884])
('occ_husb_4', [0.16140913839761342])
('occ_husb_5', [0.18777109156281066])
('occ_husb_6', [0.1940164734021257])
('rate_marriage', [-0.7031197303674775])
('age', [-0.058418128652529135])
('yrs_married', [0.10567669374160807])
('children', [0.016919829103316658])
('religious', [-0.37113511601893623])
('educ', [0.004015837061451848])


In [20]:
# evaluate the model by splitting into train and test sets
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=0)

In [21]:
model2 = LogisticRegression()
model2.fit(x_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [23]:
# predict class labels for the test set
predicted = model2.predict(x_test)
predicted

array([1., 0., 0., ..., 0., 0., 0.])

In [25]:
# generate evaluation metrics
print(metrics.accuracy_score(y_test, predicted))


0.7298429319371728


##### The accuracy is 73%, which is the same as we experienced when training and predicting on the same data.